src/v/s3/CMakeLists.txt

```diff
@@ -9,6 +9,7 @@ v_cc_library(
   DEPS
     Seastar::seastar
     v::bytes
+    v::cloud_roles
```

Comment:
would it make sense to extract the uri related helper to the http namespace? the parse_timestamp helper probably would need to be duplicated because it is aws specific.

Reply:
URI helper can be extracted to http or copied. I don't have much of an opinion here. I think that generally it's ok to have a little bit of code duplication if it helps to make things easier. It's not like URI encode function is going to be changed anytime soon. It's from the spec.

Reply:
I just realised that we use not just the helpers but several types from `s3` namespace:

1. aws region name
2. private and public keys
3. signature_v4

these are all defined in s3.

Reply:
I explored putting `cloud_roles` inside `cloud_storage` and there is still a cyclic dependency, as `cloud_roles` still requires the signature etc from s3, and the s3 client pool still requires the credential appliers.

I was able to work around it by moving the credentials and type defs to cloud_roles. So now:

1. s3 requires cloud_roles
2. cloud_storage requires s3 and cloud_roles
3. cloud_roles requires nothing

the only thing I did not like with this refactor is that I had to move the `aws_region_name` to cloud_roles. It seems like if there were a top level header only library to just define types, it would be much cleaner. Then all other libraries could import the types library and add their own dependencies.